src/java.net.http/share/classes/jdk/internal/net/http/ConnectionPool.java

```diff
@@ -524,18 +525,37 @@ private boolean contains0(HttpConnection c) {
         return false;
     }
 
-    void cleanup(HttpConnection c, Throwable error) {
+    void cleanup(HttpConnection c, long pendingData, Throwable error) {
         if (debug.on())
             debug.log("%s : ConnectionPool.cleanup(%s)",
                     String.valueOf(c.getConnectionFlow()), error);
         stateLock.lock();
+        boolean removed;
         try {
-            removeFromPool(c);
+            removed = removeFromPool(c);
             expiryList.remove(c);
         } finally {
             stateLock.unlock();
         }
-        c.close();
+        if (!removed && pendingData != 0) {
+            // this should not happen; the cleanup may have consumed
+            // some data that wasn't supposed to be consumed, so
+            // the only thing we can do is log it and close the
+            // connection.
+            if (Log.errors()) {
+                Log.logError("WARNING: CleanupTrigger triggered for" +
+                        " a connection not found in the pool: closing {0}", c.dbgString());
+            }
+            if (debug.on()) {
+                debug.log("WARNING: CleanupTrigger triggered for" +
+                        " a connection not found in the pool: closing %s", c.dbgString());
+            }
+            Throwable cause = new IOException("Unexpected cleanup triggered for non pooled connection", error);
+            cause.printStackTrace();
```

Comment:
Is the stacktrace printing intentional?

Reply:
Oh darn! That's a left over. Thanks for catching it.